File: app/src/main/java/com/smartwalkie/voiceping/MainActivity.java
package com.smartwalkie.voiceping;

import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.support.design.widget.TextInputLayout;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.Spinner;
import android.widget.Toast;

import com.smartwalkie.voicepingsdk.callbacks.DisconnectCallback;
import com.smartwalkie.voicepingsdk.exceptions.PingException;
import com.smartwalkie.voicepingsdk.listeners.AudioInterceptor;
import com.smartwalkie.voicepingsdk.listeners.AudioPlayer;
import com.smartwalkie.voicepingsdk.listeners.AudioRecorder;
import com.smartwalkie.voicepingsdk.listeners.ChannelListener;
import com.smartwalkie.voicepingsdk.listeners.OutgoingTalkCallback;
import com.smartwalkie.voicepingsdk.models.ChannelType;

import java.nio.ByteBuffer;
import java.nio.ShortBuffer;

public class MainActivity extends AppCompatActivity implements AdapterView.OnItemSelectedListener,
        ChannelListener, OutgoingTalkCallback {

    private static final String TAG = MainActivity.class.getSimpleName();

    private static final String[] CHANNEL_TYPES = { "PRIVATE", "GROUP" };

    private EditText receiverIdText;
    private Button talkButton;
    private Spinner channelTypeSpinner;
    private TextInputLayout channelInputLayout;
    private LinearLayout llOutgoingTalk;
    private ProgressBar pbOutgoingTalk;
    private LinearLayout llIncomingTalk;
    private ProgressBar pbIncomingTalk;
    private int channelType = ChannelType.PRIVATE;

    private final View.OnTouchListener touchListener = new View.OnTouchListener() {
        @Override
        public boolean onTouch(View v, MotionEvent event) {
            switch (event.getAction()) {
                case MotionEvent.ACTION_DOWN:
                    String receiverId = receiverIdText.getText().toString().trim();

                    if (receiverId == null || receiverId.isEmpty()) {
                        receiverIdText.setError(getString(R.string.error_invalid_user_id));
                        receiverIdText.requestFocus();
                        break;
                    }

                    talkButton.setText("RELEASE TO STOP");
                    talkButton.setBackgroundColor(Color.YELLOW);
                    //VoicePing.startTalking(64, ChannelType.PRIVATE);
                    VoicePingClientApp.getVoicePing().startTalking(receiverId, channelType, MainActivity.this);
                    break;
                case MotionEvent.ACTION_UP:
                    talkButton.setText("START TALKING");
                    talkButton.setBackgroundColor(Color.GREEN);
                    VoicePingClientApp.getVoicePing().stopTalking();
                    break;
                case MotionEvent.ACTION_CANCEL:
                    talkButton.setText("START TALKING");
                    talkButton.setBackgroundColor(Color.GREEN);
                    VoicePingClientApp.getVoicePing().stopTalking();
                    break;
            }
            return false;
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        channelInputLayout = (TextInputLayout) findViewById(R.id.channel_input_layout);
        channelTypeSpinner = (Spinner) findViewById(R.id.channel_type_spinner);
        ArrayAdapter<String> adapter = new ArrayAdapter<>(MainActivity.this,
                android.R.layout.simple_spinner_item, CHANNEL_TYPES);

        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        channelTypeSpinner.setAdapter(adapter);
        channelTypeSpinner.setOnItemSelectedListener(this);

        receiverIdText = (EditText) findViewById(R.id.receiver_id_text);

        talkButton = (Button) findViewById(R.id.talk_button);
        talkButton.setOnTouchListener(touchListener);

        llOutgoingTalk = (LinearLayout) findViewById(R.id.ll_outgoing_talk);
        pbOutgoingTalk = (ProgressBar) findViewById(R.id.pb_outgoing_talk);
        llIncomingTalk = (LinearLayout) findViewById(R.id.ll_incoming_talk);
        pbIncomingTalk = (ProgressBar) findViewById(R.id.pb_incoming_talk);

        llOutgoingTalk.setVisibility(View.GONE);
        llIncomingTalk.setVisibility(View.GONE);

        String userId = getIntent().getStringExtra("user_id");
        if (userId != null) setTitle("User ID: " + userId);
        talkButton.setText("START TALKING");
        talkButton.setBackgroundColor(Color.GREEN);

        VoicePingClientApp.getVoicePing().setChannelListener(this);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_disconnect:
                AlertDialog.Builder builder = new AlertDialog.Builder(this);
                builder.setTitle(R.string.text_button_disconnect)
                        .setMessage("Are you sure you want to disconnect?")
                        .setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int which) {
                                VoicePingClientApp.getVoicePing()
                                        .disconnect(new DisconnectCallback() {
                                    @Override
                                    public void onDisconnected() {
                                        Log.v(TAG, "onDisconnected...");
                                        if (!isFinishing()) {
                                            startActivity(new Intent(MainActivity.this,
                                                    LoginActivity.class));
                                            finish();
                                            Toast.makeText(MainActivity.this, "Disconnected!",
                                                    Toast.LENGTH_SHORT).show();
                                        }
                                    }
                                });
                            }
                        })
                        .setNegativeButton(android.R.string.no, new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int which) {
                                // do nothing
                            }
                        })
                        .setIcon(android.R.drawable.ic_dialog_alert)
                        .show();
                break;
            default:
                break;
        }
        return true;
    }

    // OnItemSelectedListener
    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        if (parent != channelTypeSpinner) return;
        switch (position) {
            case 0:
                receiverIdText.setHint("Receiver ID");
                channelType = ChannelType.PRIVATE;
                break;
            case 1:
                receiverIdText.setHint("Group ID");
                channelType = ChannelType.GROUP;
                break;
            default:
                break;
        }
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) {

    }

    // ChannelListener
    @Override
    public void onSubscribed(String channelId, int channelType) {

    }

    @Override
    public void onIncomingTalkStarted(AudioPlayer audioPlayer) {
        Log.d(TAG, "onIncomingTalkStarted");
        llIncomingTalk.setVisibility(View.VISIBLE);
        audioPlayer.addAudioInterceptor(new AudioInterceptor() {
            @Override
            public byte[] proceed(byte[] data) {
                ShortBuffer sb = ByteBuffer.wrap(data).asShortBuffer();
                short[] dataShortArray = new short[sb.limit()];
                sb.get(dataShortArray);
                final double amplitude = getRmsAmplitude(dataShortArray);
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        pbIncomingTalk.setProgress((int) amplitude - 7000);
                    }
                });
                return data;
            }
        });
    }

    @Override
    public void onIncomingTalkStopped() {
        Log.d(TAG, "onIncomingTalkStopped");
        llIncomingTalk.setVisibility(View.GONE);
    }

    @Override
    public void onUnsubscribed(String channelId, int channelType) {

    }

    @Override
    public void onChannelError(PingException e) {

    }

    // OutgoingTalkCallback
    @Override
    public void onOutgoingTalkStarted(AudioRecorder audioRecorder) {
        Log.d(TAG, "onOutgoingTalkStarted");
        llOutgoingTalk.setVisibility(View.VISIBLE);
        audioRecorder.addAudioInterceptor(new AudioInterceptor() {
            @Override
            public byte[] proceed(byte[] data) {
                ShortBuffer sb = ByteBuffer.wrap(data).asShortBuffer();
                short[] dataShortArray = new short[sb.limit()];
                sb.get(dataShortArray);
                final double amplitude = getRmsAmplitude(dataShortArray);
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        pbOutgoingTalk.setProgress((int) amplitude - 7000);
                    }
                });
                return data;
            }
        });
    }

    @Override
    public void onOutgoingTalkStopped() {
        Log.d(TAG, "onOutgoingTalkStopped");
        llOutgoingTalk.setVisibility(View.GONE);
    }

    @Override
    public void onOutgoingTalkError(PingException e) {
        e.printStackTrace();
        llOutgoingTalk.setVisibility(View.GONE);
        Toast.makeText(this, e.getMessage(), Toast.LENGTH_SHORT).show();
    }

    private double getRmsAmplitude(short[] dataShortArray) {
        double sum = 0;
        for (short singleShort : dataShortArray) {
            sum += singleShort * singleShort;
        }
        double meanSquare = sum / dataShortArray.length;
        return Math.sqrt(meanSquare);
    }

    private double getMaxAmplitude(short[] dataShortArray) {
        double max = Math.abs(dataShortArray[0]);
        for (short singleShort : dataShortArray) {
            if (max < singleShort) max = Math.abs(singleShort);
        }
        return max;
    }
}

File: app/src/main/java/com/smartwalkie/voicepingdemo/MainActivity.kt
package com.smartwalkie.voicepingdemo

import android.annotation.SuppressLint
import androidx.appcompat.app.AppCompatActivity
import com.smartwalkie.voicepingsdk.model.ChannelType
import android.os.Bundle
import com.smartwalkie.voicepingsdk.exception.VoicePingException
import android.content.Intent
import android.util.Log
import android.view.*
import android.widget.*
import androidx.core.content.ContextCompat
import androidx.core.widget.addTextChangedListener
import com.smartwalkie.voicepingdemo.databinding.ActivityMainBinding
import com.smartwalkie.voicepingsdk.ConnectionState
import com.smartwalkie.voicepingsdk.VoicePing
import com.smartwalkie.voicepingsdk.VoicePingButton
import com.smartwalkie.voicepingsdk.callback.ConnectCallback
import com.smartwalkie.voicepingsdk.exception.ErrorCode
import com.smartwalkie.voicepingsdk.listener.*
import com.smartwalkie.voicepingsdk.model.Channel
import java.nio.ByteBuffer

class MainActivity : AppCompatActivity(), AdapterView.OnItemSelectedListener,
    ConnectionStateListener, IncomingTalkListener {
    private val TAG = "MainActivity"
    private lateinit var binding: ActivityMainBinding

    private var mDestinationPath: String? = null
    private var mToast: Toast? = null
    private var channelType = ChannelType.GROUP
    private var disconnectConfirmationDialog: DisconnectConfirmationDialog? = null

    @SuppressLint("ClickableViewAccessibility")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val userId = MyPrefs.userId ?: ""
        val company = MyPrefs.company ?: ""
        val serverUrl = MyPrefs.serverUrl ?: ""
        if (userId.isBlank() || company.isBlank() || serverUrl.isBlank()) {
            finish()
            return
        }
        initToolbar(userId, company)
        binding.textServerUrl.text = serverUrl
        val channelTypes = arrayOf("GROUP CALL", "PRIVATE CALL")
        val adapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, channelTypes)
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        binding.spinnerChannelType.adapter = adapter
        binding.spinnerChannelType.onItemSelectedListener = this
        binding.buttonJoin.setOnClickListener { joinGroup() }
        binding.buttonLeave.setOnClickListener { leaveGroup() }
        binding.layoutGroupButtons.visibility = View.VISIBLE
        binding.buttonMute.setOnClickListener { muteChannel() }
        binding.buttonUnmute.setOnClickListener { unmuteChannel() }
        binding.layoutIncomingTalk.visibility = View.GONE
        binding.editReceiverId.addTextChangedListener {
            val receiverId = it.toString()
            binding.voicePingButton.receiverId = receiverId
            binding.voicePingButton.setButtonEnabled(receiverId.isNotBlank())
        }
        binding.voicePingButton.listener = object : VoicePingButton.Listener {
            override fun onStarted() {
                log("VoicePingButton, PTT onStarted")
            }

            override fun onStopped() {
                log("VoicePingButton, PTT onStopped")
            }

            override fun onError(errorMessage: String) {
                log("VoicePingButton, PTT error: $errorMessage")
                val receiverId = binding.editReceiverId.text.toString().trim { it <= ' ' }
                if (receiverId.isEmpty()) {
                    binding.editReceiverId.error = getString(R.string.cannot_be_blank)
                }
            }
        }
        VoicePing.setIncomingTalkListener(this)
        binding.voicePingButton.channelType = ChannelType.PRIVATE
        binding.voicePingButton.setButtonEnabled(false)
        updateConnectionState(VoicePing.getConnectionState())
        VoicePing.setConnectionStateListener(this)
        if (VoicePing.getConnectionState() == ConnectionState.DISCONNECTED) {
            VoicePing.connect(serverUrl, userId, company, object : ConnectCallback {
                override fun onConnected() {
                    // Ignored
                }

                override fun onFailed(exception: VoicePingException) {
                    // Ignored
                }
            })
        }
    }

    private fun initToolbar(userId: String, company: String) {
        supportActionBar?.title = "User ID: $userId"
        supportActionBar?.subtitle = "Company: $company"
    }

    private fun updateConnectionState(connectionState: ConnectionState) {
        log("updateConnectionState: ${connectionState.name}")
        binding.textConnectionState.text = connectionState.name
        val colorResId = when (connectionState) {
            ConnectionState.DISCONNECTED -> R.color.red
            ConnectionState.CONNECTING -> R.color.yellow
            ConnectionState.CONNECTED -> R.color.green
        }
        binding.textConnectionState.setTextColor(ContextCompat.getColor(this, colorResId))
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_open_player -> startActivity(
                PlayerActivity.generateIntent(this, mDestinationPath)
            )
            R.id.action_disconnect -> showDisconnectConfirmationDialog()
        }
        return true
    }

    // OnItemSelectedListener
    override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        if (parent !== binding.spinnerChannelType) return
        when (position) {
            0 -> {
                binding.textReceiverIdLabel.text = "Group ID"
                binding.editReceiverId.setText("${MyPrefs.company}")
                channelType = ChannelType.GROUP
                binding.layoutGroupButtons.visibility = View.VISIBLE
                binding.voicePingButton.channelType = ChannelType.GROUP
            }
            1 -> {
                binding.textReceiverIdLabel.text = "Target User ID"
                channelType = ChannelType.PRIVATE
                binding.editReceiverId.setText("@${MyPrefs.company}")
                binding.layoutGroupButtons.visibility = View.GONE
                binding.voicePingButton.channelType = ChannelType.PRIVATE
            }
        }
    }

    override fun onNothingSelected(parent: AdapterView<*>?) {}

    // ConnectionStateListener
    override fun onConnectionStateChanged(connectionState: ConnectionState) {
        runOnUiThread {
            updateConnectionState(connectionState)
        }
    }

    override fun onConnectionError(e: VoicePingException) {
        runOnUiThread {
            if (e.errorCode == ErrorCode.DUPLICATE_CONNECT) {
                if (!isFinishing) {
                    VoicePing.unmuteAll()
                    MyPrefs.clear()
                    startActivity(Intent(this, LoginActivity::class.java))
                    finish()
                }
            } else {
                showToast(e.message)
            }
        }
    }

    // IncomingTalkListener
    override fun onIncomingTalkStarted(
        audioReceiver: AudioReceiver,
        activeChannels: List<Channel>
    ) {
        log("onIncomingTalkStarted, channel: ${audioReceiver.channel.toString()}, session id: ${audioReceiver.audioSessionId}")
        // Audio processing
        Utils.enhanceLoudnessIfPossible(audioReceiver.audioSessionId, 300)
        Utils.boostBassIfPossible(audioReceiver.audioSessionId, 100.toShort())
        runOnUiThread {
            val channelType = audioReceiver.channel?.type ?: -1
            binding.layoutIncomingTalk.visibility = View.VISIBLE
            binding.textIncomingChannelType.text = ChannelType.getText(channelType)
            binding.textIncomingSenderId.text = audioReceiver.channel?.pureSenderId
        }
//        mDestinationPath = getExternalFilesDir(null).toString() + "/incoming_ptt_audio.opus"
//        audioReceiver.saveToLocal(mDestinationPath)
        audioReceiver.setInterceptorAfterDecoded(object : AudioInterceptor {
            override fun proceed(data: ByteArray, channel: Channel): ByteArray {
                val sb = ByteBuffer.wrap(data).asShortBuffer()
                val dataShortArray = ShortArray(sb.limit())
                sb[dataShortArray]
                val amplitude = Utils.getRmsAmplitude(dataShortArray)
                runOnUiThread {
                    binding.progressIncomingTalk.progress = amplitude.toInt() - 7000
                }
                return data
            }
        })
    }

    override fun onIncomingTalkStopped(
        audioMetaData: AudioMetaData,
        activeChannels: List<Channel>
    ) {
        log("onIncomingTalkStopped, channel: ${audioMetaData.channel.toString()}, download url: ${audioMetaData.downloadUrl}, active channels count: ${activeChannels.size}")
        if (activeChannels.isEmpty()) {
            runOnUiThread {
                binding.layoutIncomingTalk.visibility = View.GONE
            }
        }
    }

    override fun onIncomingTalkError(e: VoicePingException) {
        e.printStackTrace()
        runOnUiThread {
            binding.layoutIncomingTalk.visibility = View.GONE
        }
    }

    private fun joinGroup() {
        val groupId = binding.editReceiverId.text.toString().trim { it <= ' ' }
        if (groupId.isBlank()) {
            binding.editReceiverId.error = getString(R.string.cannot_be_blank)
            binding.editReceiverId.requestFocus()
            return
        }
        log("joinGroup, group ID: $groupId")
        VoicePing.joinGroup(groupId)
        showToast("Joined to $groupId")
        Utils.closeKeyboard(this, currentFocus)
    }

    private fun leaveGroup() {
        val groupId = binding.editReceiverId.text.toString().trim { it <= ' ' }
        if (groupId.isBlank()) {
            binding.editReceiverId.error = getString(R.string.cannot_be_blank)
            binding.editReceiverId.requestFocus()
            return
        }
        log("leaveGroup, group ID: $groupId")
        VoicePing.leaveGroup(groupId)
        showToast("Left from $groupId")
        Utils.closeKeyboard(this, currentFocus)
    }

    private fun muteChannel() {
        val receiverId = binding.editReceiverId.text.toString().trim { it <= ' ' }
        if (receiverId.isBlank()) return
        log("muteChannel, target ID: $receiverId, channel type: ${ChannelType.getText(channelType)}")
        VoicePing.mute(receiverId, channelType)
        showToast("Channel $receiverId muted")
    }

    private fun unmuteChannel() {
        val receiverId = binding.editReceiverId.text.toString().trim { it <= ' ' }
        if (receiverId.isBlank()) return
        log("unmuteChannel, target ID: $receiverId, channel type: ${ChannelType.getText(channelType)}")
        VoicePing.unmute(receiverId, channelType)
        showToast("Channel $receiverId unmuted")
    }

    private fun showDisconnectConfirmationDialog() {
        if (disconnectConfirmationDialog == null) {
            disconnectConfirmationDialog =
                DisconnectConfirmationDialog(this, object : DisconnectConfirmationDialog.Listener {
                    override fun onDisconnected() {
                        startActivity(Intent(this@MainActivity, LoginActivity::class.java))
                        finish()
                    }
                })
        }
        disconnectConfirmationDialog?.show()
    }

    private fun showToast(message: String?) {
        runOnUiThread {
            mToast?.cancel()
            mToast = Toast.makeText(this@MainActivity, message, Toast.LENGTH_SHORT)
            mToast?.show()
        }
    }

    private fun log(message: String) {
        Log.d(TAG, message)
    }
}